usersession/xdgopenproxy: move PortalLauncher class to own package

mardy opened the pull request:
Create a new package fox XDG desktop portals. There's some other
portal-related code in `cmd/snap/cmd_run.go`, and more will likely come in
the future, so it makes sense to have a package for all this stuff.

jhenstridge commented:
While I think it is a good idea to extract common portals functionality into a separate package, I don't think this strikes the right balance.  In particular:

1. `cmd_run.go` is a special case in that we want to minimise the overheads of launching applications, and to sanity check that the document portal is behaving the way our AppArmor and mount rules expect it to. It's not obvious that those same trade offs will make sense for other pieces of code trying to talk to xdg-document-portal. If we've got a package whose purpose is to talk to the portal services, then it's quite possible someone will come through and remove the code designed to avoid talking to the portal services as part of a refactor, ignoring `snap run`'s special requirements.
2. The `portalLauncher` type implements an intersection of the `snap userd` and xdg-desktop-portal OpenURI APIs for use by the xdg-open proxy. It's not clear it is the shape you'd pick without that constraint (e.g. why two methods on an empty struct rather than two functions?).

As an alternative, it might make sense for the `portalLauncher` type to be implemented in terms of some API provided by the `desktop/portal` package for instance. It would probably be worth driving this kind of work by extracting features as you want to reuse them, at least initially.

I'd also suggest structuring the APIs to have the caller pass in a D-Bus connection to use. This is both helpful for testing and to make it easier to control the lifetime of those connections.

mardy commented:
Thanks for the feedback, James. I'm not really trying to make a generic portal package that could be imported in other applications; my intent is just to collect the portal-related code into a separate package.

> If we've got a package whose purpose is to talk to the portal services, then it's quite possible someone will come through and remove the code designed to avoid talking to the portal services as part of a refactor, ignoring `snap run`'s special requirements.

I added a comment on the Activate method. I don't think we should do much more than this.

> I'd also suggest structuring the APIs to have the caller pass in a D-Bus connection to use. This is both helpful for testing and to make it easier to control the lifetime of those connections.

As I wrote, I'm not aiming to make a generic package for portal support, so I'm actually trying to minimize the amount of changes (at least in this PR). I've added tests for the Document portal, which were completely missing, and the `testutils.DBusTest` component seems to fit this our case well.

jhenstridge commented:
> I added a comment on the Activate method. I don't think we should do much more than this.

As I said, I think it is a mistake to move this part of the code. If you want to provide a wrapper API for for the GetMountPoint call in this new package to be called by `snap run`, that seems fine. I am saying that the handling of the `.portals-unavailable` file and sanity check for the mount point location should remain where they are.

>  I've added tests for the Document portal, which were completely missing, and the `testutils.DBusTest` component seems to fit this our case well.

That aspect of `snap run` was being tested via the `tests/main/document-portal-activation/task.yaml` spread test.

> As I wrote, I'm not aiming to make a generic package for portal support, so I'm actually trying to minimize the amount of changes (at least in this PR).

On the xdgopenproxy side, taking an interface that was private to a package and adding cross-package implementations is an increase in complexity. I'm not opposed to you moving the guts of the `portalLauncher` methods over to this new package: I'm just suggesting that you leave the type where it is, and implement its methods with calls to functions in the new package. There's very little reason to tie the two functions together like that outside of xdgopenproxy.

mardy commented:
> As I said, I think it is a mistake to move this part of the code. If you want to provide a wrapper API for for the GetMountPoint call in this new package to be called by `snap run`, that seems fine. I am saying that the handling of the `.portals-unavailable` file and sanity check for the mount point location should remain where they are.

Ah, got it! I added a commit which I think implements your suggestion, but let me know is I misunderstood.

> On the xdgopenproxy side, taking an interface that was private to a package and adding cross-package implementations is an increase in complexity. I'm not opposed to you moving the guts of the `portalLauncher` methods over to this new package: I'm just suggesting that you leave the type where it is, and implement its methods with calls to functions in the new package. There's very little reason to tie the two functions together like that outside of xdgopenproxy.

Sorry, I'm not sure I understand what you mean by "leave the type where it is": the `portalLauncher` type methods are all related to wrapping the Launcher portal APIs, so once I move them there's nothing left. If I missed something, can you please be a bit more specific on what should stay in the old `portalLauncher` type, and what I can move?

mardy commented:
After a chat with James, I think I understood his point and added a commit which should address it.

jhenstridge commented on desktop/portal/launcher.go:
```diff
@@ -0,0 +1,192 @@
+}
+
+// Launcher is a launcher that forwards the requests to xdg-desktop-portal DBus API
+type Launcher struct{}
```
As mentioned in previous comments, I don't think there is much benefit in grouping the code below as methods on a type.  We're not trying to implement xdgopenproxy's desktopLauncher interface any more, so just use functions.

mardy replied:
OK. I was actually thinking of moving the D-Bus connection to the constructor of the Launcher type, but this can also wait -- unless we plan to add more methods to the interface, I guess that removing the type is fine.

Done.

jhenstridge reviewed (changes requested):
I've left a few comments on the updated PR.

mardy commented on desktop/portal/document.go:
```diff
@@ -0,0 +1,82 @@
+	xdgRuntimeDir string
+}
+
+func (p *Document) GetUserXdgRuntimeDir() (string, error) {
+	if p.xdgRuntimeDir == "" {
+		u, err := userCurrent()
+		if err != nil {
+			return "", fmt.Errorf(i18n.G("cannot get the current user: %s"), err)
+		}
+		p.xdgRuntimeDir = filepath.Join(dirs.XdgRuntimeDirBase, u.Uid)
+	}
+	return p.xdgRuntimeDir, nil
+}
+
+func (p *Document) GetDefaultMountPoint() (string, error) {
+	xdgRuntimeDir, err := p.GetUserXdgRuntimeDir()
+	if err != nil {
+		return "", err
+	}
+
+	return filepath.Join(xdgRuntimeDir, "doc"), nil
+}
```
The second yes, it should be kept here because it belongs to this interface. The first one would be better placed in some generic type (ideally a generic one for the Desktop portal), but moving it is not trivial if we want to cache its return value. I've added a comment about it, I suggest leaving it here and eventually come back to it if we see a need to reuse it.

mardy commented on desktop/portal/document.go:
```diff
@@ -0,0 +1,82 @@
+}
+
+func (p *Document) GetMountPoint() (string, error) {
+	conn, err := dbusutil.SessionBus()
```
I initially implemented your suggestion ([in this branch](https://github.com/mardy/snapd/commit/65344898c4350f1318d941014e460db97fb64cbe)), but I didn't like it: the D-Bus connection is an implementation detail of the portals, and it doesn't sound like a good idea to expose it, if it's only for testing. I made instead another change, where one could mock the `dbusutilGetSession` method used to obtain the D-Bus session.

jhenstridge reviewed (approved):
This looks good enough.  I'd be happier if `Document.GetMountPoint` wasn't potentially creating a new D-Bus connection (as I've previously noted in comments above), but this does seem like a positive clean up over all.

MiguelPires reviewed (approved):
LGTM

mardy commented on desktop/portal/document.go:
```diff
@@ -0,0 +1,87 @@
+
+var (
+	userCurrent        = user.Current
+	osGetenv           = os.Getenv
```
Oops, removed!
